flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointStatsTracker.java

```diff
@@ -343,6 +376,62 @@ public void reportIncompleteStats(
         }
     }
 
+    public void reportInitializationStartTs(long initializationStartTs) {
+        jobInitializationMetricsBuilder =
+                Optional.of(
+                        new JobInitializationMetricsBuilder(
+                                totalNumberOfSubTasks, initializationStartTs));
+    }
+
+    public void reportInitializationMetrics(SubTaskInitializationMetrics initializationMetrics) {
+        statsReadWriteLock.lock();
+        try {
+            if (!jobInitializationMetricsBuilder.isPresent()) {
+                LOG.warn(
+                        "Attempted to report SubTaskInitializationMetrics [{}] without jobInitializationMetricsBuilder present",
+                        initializationMetrics);
+                return;
+            }
+            jobInitializationMetricsBuilder
+                    .get()
+                    .reportInitializationMetrics(initializationMetrics);
+            if (jobInitializationMetricsBuilder.get().isComplete()) {
+                traceInitializationMetrics(jobInitializationMetricsBuilder.get().build());
+            }
+        } catch (Exception ex) {
+            LOG.warn("Fail to log SubTaskInitializationMetrics[{}]", ex, initializationMetrics);
```

Comment:
When that happens, would it make sense to clear the optional to avoid further reporting work? We will never get to completion for this checkpoint anymore.